File: solutions/6.ZigZag_Conversion/zigzag-conversion.py
# Source : https://leetcode.com/problems/zigzag-conversion/
# Author : Zeming (Vantic) Zhou
# Date   : 2016-12-03

class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        sLen = len(s)
        if numRows == 1 or numRows == sLen:
            return s
        step = 2 * numRows - 2
        result = ''
        
        for i in xrange(numRows):
            for j in xrange(i, sLen, step):
                result += s[j]
                if 0 < i < numRows - 1 and j + step - 2 * i < sLen:
                    result += s[j + step - 2 * i]

        return result
             
if __name__ == '__main__':
    print Solution().convert("PAYPALISHIRING", 3)

File: solutions/6.ZigZag_Conversion/zigzag-conversion.js
// Source : https://leetcode.com/problems/zigzag-conversion/
// Author : Zeming (Vantic) Zhou
// Date   : 2016-12-03

/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */

var convert = function(s, numRows) {
  var sLength = s.length;
  if (numRows === 1 || numRows === sLength) {
    return s;
  }
  var step = 2 * numRows - 2,
      result = '';
      
  for (var i = 0; i < numRows; i++) {
    for (var j = i; j < sLength; j+=step) {
      result += s[j];
      if (0 < i  && i < numRows-1 && j + step - i * 2 < sLength) {
        result += s[j + step - i * 2];
      }
    }
  }
  
  return result;
};
